src/abstract-api-adapter.coffee

```diff
@@ -0,0 +1,103 @@
+request = require 'request'
+try
+	{ Adapter } = require 'hubot'
+catch
+	prequire = require 'parent-require'
+	{ Adapter } = prequire 'hubot'
+
+class AbstractAPIAdapter extends Adapter
+	constructor: ->
+		super
+		@interval = 1000
+		@lastReport = null
+		@allowedToPoll = true
+		throw new TypeError('Abstract class')
+
+	run: ->
+		setInterval @maybePoll, @interval
+		@emit 'connected'
+
+	maybePoll: =>
+		if @allowedToPoll
+			@allowedToPoll = false
+			@poll()
+
+	# Execute and callback a series of paged requests until we run out of pages or a filter rejects
+	getUntil: (options, each, filter, done) ->
+		request.get? options, (error, response, body) =>
+			if error or response?.statusCode isnt 200
+				if response?.statusCode in [429, 503]
+					@report { error: error, html: response?.statusCode, url: options.url }
+					# Retry if the error code indicates temporary outage
+					@getUntil options, filter, each, done
+				else
+					@robot.logger.error response?.statusCode + '-' + options.url
+					@processDone error, null, { error: error, html: response?.statusCode, url: options.url }, done
+			else
+				responseObject = JSON.parse(body)
+				results = @extractResults responseObject
+				for result in results
+					if (not filter?) or filter result
+						each result
+					else
+						@processDone null, results, { html: response?.statusCode, url: options.url }, done
+						return
+				if @extractNext responseObject
+					options.url = @extractNext responseObject
+					@getUntil options, filter, each, done
+				else
+					@processDone null, results, { html: response?.statusCode, url: options.url }, done
+
+	processDone: (error, response, report, done) =>
+		@allowedToPoll = true
+		@report report
+		done?(error, response)
+
+	extractResults: -> throw new TypeError('Abstract method')
+
+	extractNext: -> throw new TypeError('Abstract method')
+
+	report: (obj) =>
+		report = JSON.stringify(obj)
+		if @lastReport isnt report
+			@lastReport = report
+			@robot.logger.debug report
+
+	###*
+	* Given a set of ids make best effort to publish the text and send a new set of ids to the callback
+	* @param {string} text to publish
+	* @param {object} ids to use.  ids = {user, flow, thread?}
+	* @param {function} function to receive (error, new ids object).  ids = {thread, comment?}
+	###
+	postUsing: (text, ids, callback) =>
+		requestDetails = buildRequest(ids.user, ids.flow, text, ids.thread)
+		requestObject = @robot.http(requestDetails.url)
+		requestObject.header('Accept', 'application/json')
+		for key, value of requestObject.headers
+			requestObject.header(key, value)
+		# Format taken from https://github.com/github/hubot/blob/master/docs/scripting.md#making-http-calls
+		requestObject.post(JSON.stringify(requestObject.payload)) (error, headers, body) ->
+			if not error and headers.statusCode is 200
+				parseResponse(JSON.parse(body), callback)
+			else
+				callback(error ? headers, null)
```

Comment:
Calling with error makes sense, but headers is a weird fallback. It's be good behaviour for this to be an actual `Error` instance. Something like: `new Error("Received #{headers.statusCode} response: #{body}")` would be better.